Add Digital PTZ via mouse and touch

dbuezas opened the pull request:
Adds configurable Digital (local, fake, css-based) zoom and panning via mouse and touch.
The viewport setup can be persisted so it survives page reloads, and seamlessly adapts between landscape, portrait and full screen modes.

The navigation gestures are the same as in google maps and can be disabled one by one.

## For you to consider:
* ~~I enabled all gestures by default.~~
* ~~This may make scrolling difficult for people that has multiple videos in the same panel.~~
* ~~It may be more reasonable to enable only these by default~~
```
digital_ptz:
  mouse_drag_pan: true
  mouse_wheel_zoom: true
  mouse_double_click_zoom: true

  touch_pinch_zoom: true
  touch_drag_pan: true
  touch_tap_drag_zoom: true

  persist: true
```
* Persistence  is enabled by default
* I am not sure if there is any setting fro WebRTC with which all of this would be incompatible. Please let me know.
* It ended up being a fair bit of code, partially because I made extra effort in dividing into multiple classes for easier maintenance.
* To avoid silly mistakes, I actually [made this in typescript](https://github.com/dbuezas/pan-zoom-controller/blob/main/src/digital-ptz.ts) and transpiled it back to js.

---

I'm quite happy with how it feels and find it quite useful myself. I'm looking forward to your feedback :)

dbuezas commented:
Some combinations of options behave unexpectedly. I'll make a second pass tonight.

dbuezas commented:
Current defaults are:
```yaml
digital_ptz:
  mouse_drag_pan: true
  mouse_wheel_zoom: true
  mouse_double_click_zoom: true
  touch_tap_drag_zoom: true
  touch_drag_pan: false
  touch_pinch_zoom: true
  touch_double_tap_zoom: true
```

I separated the logic so it is easier to grasp: each gesture is fully independent of the rest.

dbuezas commented:
I've been testing it for a couple of days and have polished a couple of rough edges.

dbuezas commented:
I think I'm satisfied now. I made a (hopefully) last round of cleanup, simplifications and optimisations. 
Looking forward to your feedback, I'm using it quite a bit myself and it feels very natural to me. I think if feels as if it was natively supported by the video tag.

zelo66 commented:
Great work, 

1.I've been testing this for a week and it works great on my phone and in regular preview on my computer.  Unfortunately in full screen mode on PC it doesn't work for me. I tested on two computers, in chrome and opera.

2. can I use the buttons in the UI to zoom in and out?
Such as you have given in your example with button card mod.

3. can I add a button that will point and zoom in the right way to the desired location? E.g. I will set a gate position and when I press the gate button the card will automatically zoom in and move to the selected location?

dbuezas commented:
Hey @AlexxIT I assume you just haven't had time yet, but let me know if there is something about the code you'd like to be implemented differently (e.g changing defaults, explaining something with comments, ...)

I understand that it is not a very small PR, I just can't find a way to make this more compact without compromising readability or corner cases (like handling full screen video).

zelo66 commented:
> Hey @zelo66, thanks for the feedback.
> 
> 1. For it to work in full screen in desktop, you need to enable webrtc's custom ui
> 
> ```yaml
> ui: true
> ```
> 
> 2. There's no support for zooming with buttons at the moment. I could add it as a new feature after this PR is merged (if Alexey agrees)
> 3. Also no feature available to set and activate presets. At some point I want to add an option to draw a box to zoom there. If/when I do, I can add some connection to trigger that via buttons.


Of course, I have set 
` ui:true`

By it not working I meant that the image zooms in just a little. I have recorded a gif. On my main camera the image zooms in even less (there is a zoom effect which disappears after about 0.5 seconds.

I checked chrome, opera and firefox.

In Chrome and Opera zooming in full screen has the effect as in the gif. In Firefox the zoom works very well, but the `" mouse_double_click_zoom"` completely does not work.


![222 (1)](https://github.com/AlexxIT/WebRTC/assets/8418119/9d1d94bb-54e7-4941-a0f5-a7459fc1c466)

dbuezas commented:
Do you have any custom styles by any chance?

zelo66 commented:
No, everything is in the default settings.

I have now checked in one more instance of HA, plus on another computer. Same effect. It zooms in a bit, looks like the aspect ratio changes and blocks further zooming.

dbuezas commented:
I can't reproduce this so I am not sure what could be happening. Please share the yaml of the card and go2rtc stream you are using for the example. Also version of HA and OS of the pc you recorded the gif with.

zelo66 commented:
```
type: custom:webrtc-camera
url: rtsp://zephyr.rtsp.stream/movie?streamKey=19b175b07ab0c45c0d22c3fa68797415
poster: https://home-assistant.io/images/cast/splash.png
muted: true
ui: true
webrtc: true
digital_ptz:
  mouse_drag_pan: true
  mouse_wheel_zoom: true
  mouse_double_click_zoom: true
  touch_pinch_zoom: true
  touch_drag_pan: false
  touch_tap_drag_zoom: true
  persist: false
```

For this example, I am using a test stream from the website:
`https://rtsp.stream/`

HA:
`Home Assistant 2023.5.4
Supervisor 2023.06.1
Operating System 10.2
Interfejs użytkownika: 20230503.3 - latest` 

OS:
`Win 10 Home 22H2`

Of course I use the latest commit. I've also found that when I turn on the developer options the zoom works normally in full screen mode. (As long as developer mode is open).
![NOWY4 (1)](https://github.com/AlexxIT/WebRTC/assets/8418119/df9b04dd-7086-419c-a4fa-f7358bfd47c0)

dbuezas commented:
It's probably windows related. I don't have a windows machine to debug it. Do you think it is a blocker for merging?

AlexxIT commented:
Hi @dbuezas. Just had time to look at your work. Two things bother me:
1. A lot of extra external files for just one feature that only a part of users will use. Each will be downloaded to the client individually. Not the best approach in modern web development. Perhaps one file would be enough. Or even including all the code in `webrtc-camera.js`. Because `video-rtc.js` is the streaming core from go2rtc and is copied "as is". And `webrtc-camera.js` is all the extra functionality.
2. I'm not sure that all constructions will work in all browsers. Especially in horrible iOS older versions. Users of old iOS are also human beings and deserve to have normal streaming cameras. Of course, as far as Apple has allowed it. I'm not sure about constructions like `...`

dbuezas commented:
Hi @AlexxIT,

## 1. Multiple files

I think the browser will reuse the same http connection to grab all files efficiently. [I read it is only discouraged if the dependency tree is very deep](https://philipwalton.com/articles/using-native-javascript-modules-in-production-today/), and I am afraid my code will make the webrtc-camera.js too big to work comfortably with. But anyway, your repo your rules :) 

What do you prefer?
* One dedicated file with all the functionality
* All inside the webrtc-camera.js file
* A small and simple bundling script so you can modularise everything but serve a single file (if you want i can use parceljs which requires no extra config files)

## 2. Modern javascript syntax

I checked the browser versions required for each modern operator:
* `import` webrtc-camera.js already uses `import` so the lowest supported version is [Safari 11.6](https://caniuse.com/?search=import%20)
* `=>` [Safari 10](https://caniuse.com/?search=%3D%3E)
* `class` [Safari 9](https://caniuse.com/?search=%3D%3E)
* `const [Safari 11](https://caniuse.com/?search=const)
* `...` [Safari 8](https://caniuse.com/?search=Spread%20syntax)
* ``` `ES6 ${template} literals` ``` [Safari 9.1](https://caniuse.com/?search=ES6%20Template%20Literals)

So it looks like we are safe. Again, your repo, so if you just don't like a particular operator I can remove it.

---

Let me know and I'll apply your request on the weekend.
If you don't actually like the feature yourself, I can of course just make a fork for the screen pinchers like me, I'll understand that too.

AlexxIT commented:
1. I think one extra file can have all zoom functionality
2. Looks like my fears were wrong

dbuezas commented:
Hey @AlexxIT,
as requested, all the code is now in a single file `digital-ptz.js`.
I also:
* Rebased so it can be merged.
* Fixed interference between touch panning and scrolling in mobile (scroll unless zoomed), which now allows for enabling it by default
* Added a way to disable digital ptz completely via `digital_ptz: false`

dbuezas commented:
Hey Alex!
Does anything else need clarification or improvement?

AlexxIT commented:
No. I'm planning to include this in next release. Just needs some time.

AlexxIT commented:
Just as I thought - the iOS12 test did not pass.

I know it's a very old browser. But Hass somehow works in it. And WebRTC somehow works in it.

And with these innovations, the whole WebRTC card is falling down :(

dbuezas commented:
Oh, no!
Do you know exactly what failed? I can try to fix it quickly

AlexxIT commented:
I have left many comments to code

dbuezas commented:
I don't see them yet. Maybe is the review not markedbas finished?

dbuezas commented on custom_components/webrtc/www/digital-ptz.js:
```diff
@@ -0,0 +1,333 @@
+    if (o.touch_drag_pan) h.push(startTouchDragPan(gestureParam));
+    if (o.touch_pinch_zoom) h.push(startTouchPinchZoom(gestureParam));
+    this.videoEl.addEventListener("loadedmetadata", this.recomputeRects);
+    this.resizeObserver = new ResizeObserver(this.recomputeRects);
```
The problem must be the`resizeObserver = () => {}` notation. Changing it to `this.resizeObserver 
 = resizeObserver.bind(this)`

dbuezas replied:
Went for defining them inside constructor instead

AlexxIT replied:
You can create function inside observer constructor and call your function from it. This will fix first problem.

But there are some similar problems in another places.

dbuezas replied:
Exactly

dbuezas commented on custom_components/webrtc/www/digital-ptz.js:
```diff
@@ -0,0 +1,333 @@
+  let fastClicks = 0;
+  const onTouchStart = (downEvent) => {
+    const isFastClick =
+      downEvent.timeStamp - lastEvent?.timeStamp < DBL_CLICK_MS;
```
Replacing with `const isFastClick =
      lastEvent && downEvent.timeStamp - lastEvent.timeStamp < DBL_CLICK_MS;`

AlexxIT replied:
Are you sure it will work without brackets?

dbuezas replied:
Yes I am 100% certain. I actually put brackets and the formatter removed them as redundant.
I can readd them for clarity if you want

dbuezas commented:
Ok, it should work now.
I ensured the code is compatibly by finding out what ECMAScript version is compatible with iOS12 (it is ES2015) and setting the typescript transpilation target to it.

Sorry for that

dbuezas commented on custom_components/webrtc/www/webrtc-camera.js:
```diff
@@ -71,6 +72,13 @@ class WebRTCCamera extends VideoRTC {
+            this.digitalPTZ = new DigitalPTZ(
+                this.querySelector(".player"),
+                this.querySelector(".player video"),
+                { ...this.config.digital_ptz, persist_key: this.config.url }
```
I missed this file

dbuezas commented:
I applied all fixes and moved initialisation to its own member function so it is consistent with the rest of your code.

AlexxIT commented:
Looks like all works fine now. But pinch action enable full screen on iPhone. But it is not your problem :)

AlexxIT commented:
Thanks to you, I have researched ESLinter and will add it to the next release. It will help to avoid such compatibility problems.

dbuezas commented:
That's a very good idea!
And prettier is great for code formatting. If you haven't tried typescript yet, I think you'd like it too :)

AlexxIT commented:
I don't want to implement any JS build stage to my projects yet. It will unnecessary increase complexity.

dbuezas commented:
True, it does add a step. I mostly work with typescript so I'm very used to it.
For digital-ptz.js I used prettier locally, let me know if I should use a different format

dbuezas commented:
I didn't consider people using style transforms themselves. I'll try adding a wrapper element around the video and transform that one instead this way they should be able to coexist.

AlexxIT commented:
They use it only for rotation and changing aspect ratio
